contracts/oracles/composite/CompositeMultiOracle.sol

```diff
@@ -0,0 +1,147 @@
+     */
+    function setSources(bytes6[] memory bases, bytes6[] memory quotes, address[] memory sources_) external auth {
+        require(
+            bases.length == quotes.length && 
+            bases.length == sources_.length,
+            "Mismatched inputs"
+        );
+        for (uint256 i = 0; i < bases.length; i++) {
+            _setSource(bases[i], quotes[i], sources_[i]);
+        }
+    }
+
+    /**
+     * @notice Set or reset an price path
+     */
+    function setPath(bytes6 base, bytes6 quote, bytes6[] memory path) external auth {
+        _setPath(base, quote, path);
+    }
+
+    /**
+     * @notice Set or reset a number of price paths
+     */
+    function setPaths(bytes6[] memory bases, bytes6[] memory quotes, bytes6[][] memory paths_) external auth {
+        require(
+            bases.length == quotes.length && 
+            bases.length == paths_.length,
+            "Mismatched inputs"
+        );
+        for (uint256 i = 0; i < bases.length; i++) {
+            _setPath(bases[i], quotes[i], paths_[i]);
+        }
+    }
+
+    /**
+     * @notice Retrieve the value of the amount at the latest oracle price.
+     * @return value
+     */
+    function peek(bytes32 base, bytes32 quote, uint256 amount)
+        external view virtual override
+        returns (uint256 value, uint256 updateTime)
+    {
+        uint256 price;
+        (price, updateTime) = _peek(base.b6(), quote.b6());
+        value = price * amount / 1e18;
+    }
+
+    /**
+     * @notice Retrieve the value of the amount at the latest oracle price.. Same as `peek` for this oracle.
+     * @return value
+     */
+    function get(bytes32 base, bytes32 quote, uint256 amount)
+        external virtual override
+        returns (uint256 value, uint256 updateTime)
+    {
+        uint256 price;
+        (price, updateTime) = _peek(base.b6(), quote.b6());
+        value = price * amount / 1e18;
+    }
+
+    function _peek(bytes6 base, bytes6 quote) private view returns (uint price, uint updateTime) {
+        bytes6[] memory path = paths[base][quote];
+        if (path.length > 0) {
+            bytes32 base_ = base;
+            for (uint256 p = 0; p < path.length; p++) {
+                Source memory source = sources[base][quote];
+                require (source.source != address(0), "Source not found");
+                uint256 price_;
+                uint256 updateTime_;
+                (price_, updateTime_) = IOracle(source.source).peek(base_, quote, 10 ** source.decimals);   // Get price for one unit
+                price = price * price_ / (10 ** source.decimals);                                           // Fixed point according to decimals
+                updateTime = (updateTime > updateTime_) ? updateTime : updateTime_;                         // Take the oldest update time
+                base_ = path[p];
+            }
+        }
+    }
+
+    function _get(bytes6 base, bytes6 quote) private returns (uint price, uint updateTime) {
```

Comment:
Where is this used?

Reply:
Some oracles have both transactional and non-transactional price feeds. For example, taking a cToken as a source for a price feed, you can use `exchangeRateCurrent` or `exchangeRateStored`.

The Cauldron always retrieves prices using transactional feeds (`get`), although then the rate and chi oracle uses the non-transactional function (`_peek`) under the hood to save gas. The Chainlink spot oracles also converge to the non-transactional function (`_peek`).

The only case where the Cauldron ends using a transaction price feed is when using a cToken as a collateral, because we want to end using `exchangeRateCurrent` in that case.